### ISSUE TITLE
Getting `Unresolvable dependency` exception

### ISSUE THREAD

**Issue by aik099** [rcrowe/TwigBridge#150]
I'm using https://github.com/barryvdh/laravel-ide-helper to generate auto-complete for PhpStorm. I guess it's invoking all services providers to work. As a result I'm getting following exception:

```
Exception: Cannot instantiate Twig extension 'TwigBridge\Extension\Laravel\Form': Unresolvable dependency resolving [Parameter #2 [ <required> $csrfToken ]] in class Illuminate\Html\FormBuilder
Skipping \TwigBridge\Facade\Twig.
A new helper file was written to _ide_helper.php
```

I guess this happens because the mentioned `TwigBridge\Extension\Laravel\Form` class has dependency on FormBuilder (see https://github.com/rcrowe/TwigBridge/blob/master/src/Extension/Laravel/Form.php#L34), which happens to have 3rd required argument called `$csrfToken`: https://github.com/laravel/framework/blob/4.2/src/Illuminate/Html/FormBuilder.php#L82

Used Laravel Version: 4.3

Related issue: barryvdh/laravel-ide-helper#119

**Comment by jeanbelhache** [comment]
Since the commit of 12 march 2015, on the branch 0,7 the register function should be :

public function register()
    {
        $this->app->bind('Collective\Html\FormBuilder', function () {

            return new \Collective\Html\FormBuilder(
                $this->app->make('Collective\Html\HtmlBuilder'),
                $this->app->make('Illuminate\Routing\UrlGenerator'),
                csrf_token()
            );

        });
    }

**Comment by barryvdh** [comment]
Can you submit a PR?

**Comment by mehrdd** [comment]
i made an update and :
```
InvalidArgumentException in ServiceProvider.php line 194:
Cannot instantiate Twig extension 'TwigBridge\Extension\Laravel\Html': Class Collective\Html\HtmlBuilder does not exist
```

**Comment by jeanbelhache** [comment]
Hi mehrdd,

you have to include  "laravelcollective/html": "~5.0"  into your composer.

see http://laravelcollective.com/docs/5.0/html

**Comment by MyIgel** [comment]
@jeanbelhache With laravelcollective/html also an InvalidArgumentException is thrown: ```Cannot instantiate Twig extension 'TwigBridge\Extension\Laravel\Form': Unresolvable dependency resolving [Parameter #2 [ <required> $csrfToken ]] in class Collective\Html\FormBuilder```

**Comment by jeanbelhache** [comment]
@MyIgel  Yes, i mentioned that issue in a pull request https://github.com/rcrowe/TwigBridge/pull/184

**Comment by enchance** [comment]
@MyIgel I followed your directions and modified `register()` in the ServiceProvider.php file but I'm getting a __Class twig.extension does not exist__ error. Did I miss anything? I'm using L5.

**Comment by amosmos** [comment]
Hi @aik099, Seemes like you're the expert on this issue so out of curiosity I would be happy if you could explain to me what happened here.

1. It seems like the problem was with TwigBridge type hinting FormBuilder class, while the FormBuilder class constructor requires the $csrf_token paramater. But if I understand correctly type hinting works by using the name that you bind to the service container and not the actual class name. And since the htmlServiceProvider is binding the formBuilder to the name 'form' - wasn't TwigBridge needed to type hint `form $form` instead of `FormBuilder $form` to solve the issue?

2. Even if aliasing 'form' to 'FormBuilder' class actually somehow solved TwigBridge issue with constructing a FormBuilder, why wasn't adding this alias to the config/app.aliases solved it? Why did it had to be aliased in the service provider?

Hope my questions made sense,
Thanks!

**Comment by amosmos** [comment]
@aik099 

The more I think about it I just don't understand why do TwigBridge's type hinting was made directly to the FormBuilder class and not to the name that got bind for it ("form").

It just seems to me that the addition to the illuminate/html package was not needed. Yes it did solved TwigBridge issue (which I actually don't even understand how as it still uses the "form" alias), but this should have been fixed on TwigBridge side.

But since so many experts were involved in this issue, I probably got something wrong and I just really wonder what it is....

Please let me know what you think,
Thanks again!

**Comment by aik099** [comment]
Sorry, but I'm not using Laravel anymore so I'm not aware if this issue was fixed or not.

**Comment by amosmos** [comment]
Got it :-)

Yes, It was fixed. But it was fix with the addition of a manual alias to the html class, while in my opinion the fix should have been on the TwigBridge class.

I'm just curious where I'm wrong in thinking the applied fix is not the right one....

Thanks!

**Comment by MyIgel** [comment]
@enchance  the fix was to add a *app/Providers/FormServiceProvider.php* with the following content:
```php
<?php namespace AppNamespace\Providers;

use Collective\Html\FormBuilder;
use Illuminate\Support\ServiceProvider;

/**
 * Class FormServiceProvider
 *
 * Fix for https://github.com/rcrowe/TwigBridge/issues/150
 *
 * @package AppNamespace\Providers
 */
class FormServiceProvider extends ServiceProvider
{
    /**
     * Register the service provider.
     *
     * @return void
     */
    public function register()
    {
        $this->app->bindIf('Collective\Html\FormBuilder', function () {
            return new FormBuilder(
                $this->app->make('Collective\Html\HtmlBuilder'),
                $this->app->make('Illuminate\Routing\UrlGenerator'),
                csrf_token()
            );
        });
    }
}
```
